// graphql/schemas/boardSchema.js
const graphql = require('graphql')
const { GraphQLObjectType, GraphQLInt, GraphQLString, GraphQLList } = graphql
const boardMongoSchema = require('../mongoDBSchema/boardSchema.js')
const authorMongoSchema = require('../mongoDBSchema/userSchema.js')
const { UserType } = require('./userSchema.js')

const boardSchema = new GraphQLObjectType({
  name: 'boardSchema',
  fields: () => ({
    boardId: ({ type: GraphQLString }),
    boardName: ({ type: GraphQLString }),
    boardScore: ({ type: GraphQLInt }),
    boardCreator: ({ type: UserType,
      resolve (parents, args) {
        return UserType.findById(parents.authorId)
      }

    })
  })
})

const boards = {
  type: GraphQLList(boardSchema),
  args: {
    id: { type: GraphQLString },
    boardId: ({ type: GraphQLString }),
    boardName: ({ type: GraphQLString }),
    boardScore: ({ type: GraphQLInt }),
    boardCreator: ({ type: GraphQLString }) },
  async resolve (parent, args) {
    var result = await boardMongoSchema.find({})
    console.log(result)
    return result
  }
}

module.exports = boards
