// graphql/schemas/topicSchema.js
const graphql = require('graphql')
const { GraphQLObjectType, GraphQLInt, GraphQLString, GraphQLSchema } = graphql
const Topic = require('../mongoDBSchema/topicSchema')
const topicType = new GraphQLObjectType({
  'name': 'Topic',
  fields: () => ({
    id: { type: GraphQLInt },
    topicTitle: { type: GraphQLString },
    topicContent: { type: GraphQLString },
    topicTimeStamp: { type: GraphQLString }
  })
})

const mutation = new GraphQLObjectType({
  'name': 'topicMutation',
  fields: { addTopic: {
    type: topicType,
    fields: {
      topicTitle: { type: GraphQLString },
      topicContent: { type: GraphQLString },
      topicTimeStamp: { type: GraphQLString }
    },
    resolve (parent, args) {
      let topic = new Topic({
        topicTitle: args.topicTitle,
        topicContent: args.topicContent,
        topicTimeStamp: args.topicTimeStamp
      })
      return topic.save()
    }

  } }

})

module.exports = new GraphQLSchema({
  query: topicType,
  mutation: mutation

})
