// graphql/schemas/topicSchema.js
const graphql = require('graphql')
const { GraphQLObjectType, GraphQLInt, GraphQLString, GraphQLList } = graphql
const topicSchema = require('../mongoDBSchema/topicSchema.js')
const topicType = new GraphQLObjectType({
  'name': 'topic',
  fields: () => ({
    topicId: { type: GraphQLInt },
    topicName: { type: GraphQLString },
    topicContent: { type: GraphQLString },
    topicTimeStamp: { type: GraphQLString },
    topicBoardName: { type: GraphQLString }
  })
})

const addTopicMutation = {
  type: topicType,
  args: {
    topicName: { type: GraphQLString },
    topicContent: { type: GraphQLString },
    topicTimeStamp: { type: GraphQLString },
    topicBoardName: { type: GraphQLString }
  },
  resolve (parent, args) {
    let topic = new topicSchema({
      topicName: args.topicName,
      topicContent: args.topicContent,
      topicTimeStamp: args.topicTimeStamp,
      topicBoardName: args.topicBoardName
    })
    return topic.save()
  } }

const loadTopicMutation = {
  type: GraphQLList(topicType),
  args: {
    topicBoardName: { type: GraphQLString }

  },
  async resolve (parent, args) {
    console.log('result')
    console.log(args.topicBoardName)
    // 'topicBoardName': args.topicBoardName
    let topicQuery = topicSchema.find({})

    topicQuery.select('topicName topicContent')

    let result = await topicQuery.exec()
    console.log(result)
    return result
  }

}

module.exports = { addTopicMutation, loadTopicMutation }
